nexus/db-queries/src/db/queries/sled_reservation.rs

```diff
@@ -0,0 +1,385 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Implementation of queries for affinity groups
+
+use crate::db::model::AffinityPolicyEnum;
+use crate::db::model::Resources;
+use crate::db::model::SledResourceVmm;
+use crate::db::raw_query_builder::QueryBuilder;
+use crate::db::raw_query_builder::TypedSqlQuery;
+use diesel::sql_types;
+use omicron_uuid_kinds::GenericUuid;
+use omicron_uuid_kinds::InstanceUuid;
+
+/// Return all possible Sleds where we might perform allocation
+///
+/// The rows returned by this CTE indicate:
+///
+/// - The Sled which we're considering
+/// - A bool indicating whether the allocation fits
+/// - Affinity Policy
+/// - Anti-Affinity Policy
+///
+/// Generally, we'd like to only return sleds where an allocation of [Resources]
+/// could fit. However, we also need to observe all affinity groups, in case
+/// affinity group policy forces us to allocate on a sled where insufficient
+/// space exists.
+///
+/// Note that we don't bother checking if the VMM has already been provisioned,
+/// we're just searching for spots where it might fit.
+pub fn sled_find_targets_query(
+    instance_id: InstanceUuid,
+    resources: &Resources,
+) -> TypedSqlQuery<(
+    sql_types::Uuid,
+    sql_types::Bool,
+    sql_types::Nullable<AffinityPolicyEnum>,
+    sql_types::Nullable<AffinityPolicyEnum>,
+)> {
+    QueryBuilder::new().sql("
+        WITH sled_targets AS (
+            SELECT sled.id as sled_id
+            FROM sled
+            LEFT JOIN sled_resource_vmm
+            ON sled_resource_vmm.sled_id = sled.id
+            WHERE
+                sled.time_deleted IS NULL AND
+                sled.sled_policy = 'in_service' AND
+                sled.sled_state = 'active'
+            GROUP BY sled.id
+            HAVING
+                COALESCE(SUM(CAST(sled_resource_vmm.hardware_threads AS INT8)), 0) + "
```

Comment:
the `COALESCE(..., 0)` here is necessary because there might be no rows matching the WHERE clause at all, right?

Reply:
Yeah, this whole query is `sled LEFT JOIN (implied outer) sled_resource_vmm`, so the `sled_resource_vmm` records might be null.

I believe that `SUM(NULL, ...)` returns NULL, so the coalesce calls here basically say "if there are no `sled_resource_vmm` records on a a sled, treat this whole value as zero".